File: src/main/java/com/example/accountbook/model/Statement.java
package com.example.accountbook.model;

import org.hibernate.annotations.ColumnDefault;
import org.hibernate.annotations.DynamicInsert;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.time.LocalDateTime;

@Entity
@DynamicInsert
@Table(indexes = {
        @Index(name = "idx_statement_account_card", columnList = "account_card_id"),
        @Index(name = "idx_statement_date", columnList = "date"),
})
@EntityListeners({AuditingEntityListener.class})
public class Statement {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(optional = false)
    private Category category;

    @NotNull
    private String name;

    @NotNull
    private int amount;

    @ColumnDefault("0")
    private int discount;

    @ManyToOne(optional = false)
    private AccountCard accountCard;

    @ColumnDefault("false")
    private boolean isAlert;

    @NotNull
    private LocalDateTime date;

    @CreatedDate
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;

    public Long getId() {
        return id;
    }

    public Category getCategory() {
        return category;
    }

    public void setCategory(Category category) {
        this.category = category;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getAmount() {
        return amount;
    }

    public void setAmount(int amount) {
        this.amount = amount;
    }

    public int getDiscount() {
        return discount;
    }

    public void setDiscount(int discount) {
        this.discount = discount;
    }

    public AccountCard getAccountCard() {
        return accountCard;
    }

    public void setAccountCard(AccountCard accountCard) {
        this.accountCard = accountCard;
    }

    public LocalDateTime getDate() {
        return date;
    }

    public void setDate(LocalDateTime date) {
        this.date = date;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public boolean isAlert() {
        return isAlert;
    }

    public void setAlert(boolean alert) {
        isAlert = alert;
    }
}

File: src/main/java/com/example/accountbook/model/Statement.kt
package com.example.accountbook.model

import org.hibernate.annotations.ColumnDefault
import org.hibernate.annotations.DynamicInsert
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.time.LocalDateTime
import javax.persistence.*
import javax.validation.constraints.NotNull

@Entity
@DynamicInsert
@Table(
    indexes = [Index(
        name = "idx_statement_account_card",
        columnList = "account_card_id"
    ), Index(name = "idx_statement_date", columnList = "date")]
)
@EntityListeners(
    AuditingEntityListener::class
)
data class Statement(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long,
    @ManyToOne(optional = false)
    var category: Category,
    var name: @NotNull String,
    var amount: @NotNull Int = 0,

    @ColumnDefault("0")
    var discount: Int = 0,

    @ColumnDefault("0")
    var saving: Int = 0,

    @ManyToOne(optional = false)
    var accountCard: AccountCard,

    @ColumnDefault("false")
    var isAlert: Boolean = false,

    var date: @NotNull LocalDateTime,

    @CreatedDate
    var createdAt: LocalDateTime,

    @LastModifiedDate
    var updatedAt: LocalDateTime,
    var description: String?,
) {
    @PrePersist
    fun prePersist() {
        if (category.type != Category.TYPE_INCOME) {
            amount *= -1
        }
    }
}
